rllib/utils/exploration/random_encoder.py

```diff
@@ -0,0 +1,338 @@
+    dist = np.linalg.norm(
+        obs_embeds_[:, None, :] - obs_embeds_[None, :, :], axis=-1)
+    return dist.argsort(axis=-1)[:, :k_nn][:, -1]
+
+
+class RE3UpdateCallbacks():
+    """Update input callbacks to mutate batch with states entropy rewards."""
+
+    _step = 0
+
+    def __init__(self,
+                 *args,
+                 embeds_dim: int = 128,
+                 k_nn: int = 50,
+                 beta: float = 0.1,
+                 rho: float = 0.0001,
+                 beta_schedule: str = "constant",
+                 **kwargs):
+        self.embeds_dim = embeds_dim
+        self.k_nn = k_nn
+        self.beta = beta
+        self.rho = rho
+        self.beta_schedule = beta_schedule
+        self._rms = MovingMeanStd()
+        super().__init__(*args, **kwargs)
+
+    def on_learn_on_batch(
+            self,
+            *,
+            policy: "Policy",
+            train_batch: SampleBatch,
+            result: dict,
+            **kwargs,
+    ):
+        # import Postprocessing here to avoid circular import issue
+        from ray.rllib.evaluation.postprocessing import Postprocessing
+
+        states_entropy = compute_states_entropy(
+            train_batch[SampleBatch.OBS_EMBEDS], self.embeds_dim, self.k_nn)
+        states_entropy = update_beta(
+            self.beta_schedule, self.beta, self.rho,
+            RE3UpdateCallbacks._step) * np.reshape(
+                self._rms(states_entropy),
+                train_batch[SampleBatch.OBS_EMBEDS].shape[:-1],
+            )
+        train_batch[SampleBatch.REWARDS] = (
+            train_batch[SampleBatch.REWARDS] + states_entropy)
+        if Postprocessing.ADVANTAGES in train_batch:
+            train_batch[Postprocessing.ADVANTAGES] = (
+                train_batch[Postprocessing.ADVANTAGES] + states_entropy)
+            train_batch[Postprocessing.VALUE_TARGETS] = (
+                train_batch[Postprocessing.VALUE_TARGETS] + states_entropy)
+        super().on_learn_on_batch(
+            policy=policy, train_batch=train_batch, result=result, **kwargs)
+
+    def on_train_result(self, *, trainer, result: dict, **kwargs) -> None:
+        RE3UpdateCallbacks._step = result["training_iteration"]
+        super().on_train_result(trainer=trainer, result=result, **kwargs)
+
+
+class RE3(Exploration):
+    """Random Encoder for Efficient Exploration.
+
+    Implementation of:
+    [1] State entropy maximization with random encoders for efficient
+    exploration. Seo, Chen, Shin, Lee, Abbeel, & Lee, (2021).
+    arXiv preprint arXiv:2102.09430.
+
+    Estimates state entropy using a particle-based k-nearest neighbors (k-NN)
+    estimator in the latent space. The state's latent representation is
+    calculated using an encoder with randomly initialized parameters.
+
+    The entropy of a state is considered as intrinsic reward and added to the
+    environment's extrinsic reward for policy optimization.
+    """
```

Comment:
should we mention here that entropy is only calculated per batch, and does not take the distribution of the entire replay buffer into consideration?
basically one of your responses, we just put it here.